### PR TITLE
Mollie subscription ept

### PR DESCRIPTION
**Pull request by dhyeyp-emipro**
[MOD] Modified code as per given feedback

### CODE REVIEW

File: mollie_subscription_ept/controllers/__init__.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
 
 from . import main
-from . import payment
+# from . import payment
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/controllers/__init__.py]: Remove this line altogether

File: mollie_subscription_ept/controllers/main.py [hunk 2]
```diff
@@ -57,37 +55,25 @@ def cart_update_json(self, product_id, line_id=None, add_qty=None, set_qty=None,
...
+                                                           line_id=line_id,
+                                                           add_qty=add_qty,
+                                                           set_qty=set_qty,
+                                                           display=display, **kw)
+
+        self._set_session(values)
+
+        return values
+
+class MolliePayment(WebsiteSale):
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/controllers/main.py]: No need to create a separate class for this

File: mollie_subscription_ept/controllers/portal.py [hunk 3]
```diff
@@ -47,7 +45,6 @@ def portal_my_subscription(self, page=1, date_begin=None, date_end=None, sortby=
                              step=self._items_per_page)
         # content according to pager and archive selected
         orders = sale_order.search(domain, order=sort_order, limit=self._items_per_page, offset=pager['offset'])
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/controllers/portal.py]: This can be changed to request.env["sale.order"] as its being used only once in the entire function

File: mollie_subscription_ept/models/sale_order.py [hunk 4]
```diff
@@ -7,19 +7,12 @@ class SaleOrder(models.Model):
...
+        return self.order_line.product_id.filtered(lambda product: product.is_mollie_subscription)
 
     def _public_user(self):
         public_partner_id = self.env.ref("base.public_user") and self.env.ref("base.public_user").partner_id or False
         current_partner_id = self.partner_id or False
-        if public_partner_id.id == current_partner_id.id:
-            return True
-        else:
-            return False
+        return True if public_partner_id.id == current_partner_id.id else False
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/models/sale_order.py]: Change this to "return True if public_partner_id == current_partner_id else False"

**Review by bvanjeelharia — COMMENTED** [review]
Please look into the new comments and no changes are made for python dependencies in manifest file

File: mollie_subscription_ept/controllers/main.py [hunk 5]
```diff
@@ -18,35 +18,6 @@ def _set_session(self, so_obj):
         request.session['show_wishlist_warn'] = True if so_obj.get('show_wishlist_warning') else False
         request.session['show_in_cart_warn'] = True if so_obj.get('show_in_cart_warning') else False
 
-    @http.route(['/shop/cart/update'], type='http', auth="public", methods=['POST'], website=True)
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/controllers/main.py]: Why was this removed?

Comment (dhyeyp-emipro) [reply mollie_subscription_ept/controllers/main.py]: There is no need for more.

Comment (bvanjeelharia) [reply mollie_subscription_ept/controllers/main.py]: Ok fine

**Comment by bvanjeelharia** [comment]
Also, changes requested in the manifest file have not been done

**Comment by dhyeyp-emipro** [comment]
It is necessary to include the Python dependencies line. Because Python dependencies must be configured on the system when installing this module. If not, then the module will not be able to request the API.

**Comment by bvanjeelharia** [comment]
'external_dependencies': {
        'python': ['mollie-api-python']
    },

it needs to be replaced by this piece of code

File: mollie_subscription_ept/models/mollie_subscription.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        return start_date.strftime("%Y-%m-%d")
+
+    def _mollie_create_subscription(self, transaction_obj):
+        log_sudo = self.env['mollie.log'].sudo()
+        log_sudo.add_log("Start Creating Subscription", "Start Creating Subscription")
+        mollie = self.env.ref("payment.payment_acquirer_mollie")
+        mollie_client = mollie._api_mollie_get_client()
+        sale_order_obj = transaction_obj.sale_order_ids
+        payment_data = transaction_obj.acquirer_id._api_mollie_get_payment_data(transaction_obj.acquirer_reference)
+        for product in sale_order_obj.order_line.product_id:
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/models/mollie_subscription.py]: use filtered here to get lines having is_mollie_subscription set to True

File: mollie_subscription_ept/models/payment_transaction.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            order_source = self.sale_order_ids[0]
+            partner_obj = order_source.partner_id
+        if partner_obj and partner_obj.mollie_customer_id:
+            mollie_customer_id = partner_obj.mollie_customer_id
+        elif partner_obj:
+            customer_id_data = self.acquirer_id._api_mollie_create_customer_id(partner_obj)
+            if customer_id_data and customer_id_data.get('id'):
+                mollie_customer_id = customer_id_data.get('id')
+                partner_obj.write({'mollie_customer_id': mollie_customer_id})
+        return mollie_customer_id
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/models/payment_transaction.py:22]: Can be modified as 
    def _get_transaction_customer_id(self):
        mollie_customer_id = False
        if self.sale_order_ids:
            partner_obj = self.sale_order_ids[0].partner_id
            if partner_obj.mollie_customer_id:
                mollie_customer_id = partner_obj.mollie_customer_id
            else:
                customer_id_data = self.acquirer_id._api_mollie_create_customer_id(partner_obj)
                if customer_id_data and customer_id_data.get('id'):
                    mollie_customer_id = customer_id_data.get('id')
                    partner_obj.write({'mollie_customer_id': mollie_customer_id})
        return mollie_customer_id

File: mollie_subscription_ept/models/sale_order.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+            if self._check_subs_in_cart() and not product_obj.is_mollie_subscription:
+                return {'show_in_cart_warning': True}
+
+            values = super(SaleOrder, self)._cart_update(product_id, line_id, add_qty, set_qty, **kwargs)
+
+            # If any non-subscription products available in the cart, move those products to the wishlist
+            if product_obj.is_mollie_subscription:
+                order_line = self.order_line.filtered(lambda x: not x.product_id.is_mollie_subscription)
+                for line in order_line:
+                    if not line.product_id._is_in_wishlist() and not self._public_user():
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/models/sale_order.py]: Replace this section
                order_line = self.order_line.filtered(lambda x: not x.product_id.is_mollie_subscription)
                is_public_user = self._public_user()
                for line in order_line:
                    if not line.product_id._is_in_wishlist() and not is_public_user:

**Review by bvanjeelharia — COMMENTED** [review]
There should be a separate Main Menu for Mollie, it shouldn't show up inside Sales.
The access rights should be changed as well.Right now , every user can access everything
If no credit card/ideal payment method is setup, it shows as blank for pay with section, a message should be displayed like 'No payment methods setup.Contact us @...'
Screenshot from 2023-03-15 15-51-23

**Comment by dhyeyp-emipro** [comment]
Modified code as per given feedback.

**Review by bvanjeelharia — APPROVED** [review]

**Review by bvanjeelharia — APPROVED** [review]

**Review by bvanjeelharia — COMMENTED** [review]
Can you please update changes so that the payment page shows the appropriate methods when making the payment for subscriptions through the payment link generated from sale order page?

**Comment by bvanjeelharia** [comment]
Yes, working fine @dhyeyp-emipro

File: mollie_subscription_ept/views/mollie_subscription_view.xml [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+            <tree string="Subscriptions" default_order='create_date desc'>
+                <field name="name"/>
+                <field name="customerId"/>
+                <field name="partner_id"/>
+                <field name="product_id"/>
+                <field name="description"/>
+                <field name="amount"/>
+                <field name="startDate"/>
+                <field name="nextPaymentDate"/>
+                <field name="status"/>
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/views/mollie_subscription_view.xml:17]: Please add <field name="create_date" invisible="1"/> as its causing issues on opening contacts for contacts having subscriptions

**Review by bvanjeelharia — APPROVED** [review]

File: mollie_subscription_ept/views/mollie_subscription_view.xml (hunk 9 above)
Comment (dhyeyp-emipro) [reply mollie_subscription_ept/views/mollie_subscription_view.xml:17]: I've already taken these changes please see this: https://tinyurl.com/24lc4w36

File: mollie_subscription_ept/views/payment_mollie_templates.xml [hunk 10]
```diff
@@ -22,4 +23,22 @@
             </div>
         </xpath>
     </template>
+
+    <template id="mollie_payment_checkout"
+              inherit_id="payment_mollie_official.mollie_payment_checkout">
+
+        <xpath expr="//form/div[hasclass('card')]" position="before">
+            <t t-if="not order">
+                <t t-set="order" t-value="request.env['sale.order'].browse(sale_order_id)"/>
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/views/payment_mollie_templates.xml]: add sudo() here as its causing issues. request.env['sale.order'].sudo()

File: mollie_subscription_ept/views/payment_mollie_templates.xml [hunk 11]
```diff
@@ -22,4 +23,22 @@
...
+            </t>
+            <t t-set="subscription_product"
+               t-value="order.order_line.product_id.filtered(lambda product: product.is_mollie_subscription)"/>
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/views/payment_mollie_templates.xml]: add sudo() here as its causing issues.order.sudo().order_line

**Comment by dhyeyp-emipro** [comment]
Changes uploaded, Please verify it from your end.

**Comment by bvanjeelharia** [comment]
Yes, seems fine

**Comment by bvanjeelharia** [comment]
Still facing issues, which is probably happening because of mollie notify(webhook) is working fine on the instance.The system is trying to process the data twice I believe.

Here are the logs:
2023-04-24 14:10:18,927 76 INFO nwb odoo.addons.payment_mollie.controllers.main: Received Mollie notify data:
{'id': 'tr_qYgFfEzY7U', 'ref': 'S00467'} 
2023-04-24 14:10:21,321 76 INFO nwb werkzeug: 87.233.217.243 - - [24/Apr/2023 14:10:21] "POST /payment/mollie/notify?ref=S00467 HTTP/1.0" 200 - 386 0.278 2.127
2023-04-24 14:10:21,420 69 INFO nwb odoo.addons.base.models.ir_cron: Starting job `automatic invoicing: send ready invoice`. 
2023-04-24 14:10:21,429 69 INFO nwb odoo.addons.base.models.ir_cron: Job `automatic invoicing: send ready invoice` done. 
2023-04-24 14:10:22,408 78 INFO nwb odoo.addons.payment_mollie.controllers.main: Received Mollie return data:
{'ref': 'S00467'} 
2023-04-24 14:10:22,947 78 INFO nwb odoo.addons.payment.models.payment_transaction: tried to write tx state with same value (ref: S00467, state: done) 
2023-04-24 14:10:22,955 78 INFO nwb werkzeug: 103.183.54.203 - - [24/Apr/2023 14:10:22] "GET /payment/mollie/return?ref=S00467 HTTP/1.0" 303 - 21 0.013 0.537
2023-04-24 14:10:23,381 64 INFO nwb werkzeug: 103.183.54.203 - - [24/Apr/2023 14:10:23] "GET /payment/status HTTP/1.0" 200 - 16 0.013 0.038
2023-04-24 14:10:23,646 78 INFO nwb werkzeug: 103.183.54.203 - - [24/Apr/2023 14:10:23] "GET /base_fontawesome/static/lib/fontawesome-5.13.0/webfonts/fa-solid-400.woff2 HTTP/1.0" 404 - 2 0.001 0.004
2023-04-24 14:10:23,648 76 INFO nwb werkzeug: 103.183.54.203 - - [24/Apr/2023 14:10:23] "GET /base_fontawesome/static/lib/fontawesome-5.13.0/webfonts/fa-regular-400.woff2 HTTP/1.0" 404 - 2 0.002 0.006
2023-04-24 14:10:23,658 61 INFO nwb werkzeug: 103.183.54.203 - - [24/Apr/2023 14:10:23] "GET /base_fontawesome/static/lib/fontawesome-5.13.0/webfonts/fa-brands-400.woff2 HTTP/1.0" 404 - 2 0.001 0.006
2023-04-24 14:10:23,950 61 INFO nwb werkzeug: 103.183.54.203 - - [24/Apr/2023 14:10:23] "GET /web/webclient/qweb/1682345423816?bundle=web.assets_frontend HTTP/1.0" 200 - 16 0.008 0.053
2023-04-24 14:10:24,348 61 INFO nwb odoo.http: <function odoo.addons.payment.controllers.post_processing.poll_status> called ignoring args <csrf_token=208074e1143a76b95dae92fd8984614e2a10c7e6o1713881423> 
2023-04-24 14:10:24,369 61 ERROR nwb odoo.addons.payment.controllers.post_processing: encountered an error while post-processing transactions with ids 356:
There is nothing to invoice!

Reason(s) of this behavior could be:
- You should deliver your products before invoicing them.
- You should modify the invoicing policy of your product: Open the product, go to the "Sales" tab and modify invoicing policy from "delivered quantities" to "ordered quantities". For Services, you should modify the Service Invoicing Policy to 'Prepaid'. 
Traceback (most recent call last):
  File "/odoo/src/addons/payment/controllers/post_processing.py", line 81, in poll_status
    txs_to_post_process._finalize_post_processing()
  File "/odoo/src/addons/website_payment/models/payment_transaction.py", line 13, in _finalize_post_processing
    super()._finalize_post_processing()
  File "/odoo/src/addons/payment/models/payment_transaction.py", line 859, in _finalize_post_processing
    self._reconcile_after_done()
  File "/odoo/src/addons/sale/models/payment_transaction.py", line 102, in _reconcile_after_done
    self._invoice_sale_orders()
  File "/odoo/src/addons/sale/models/payment_transaction.py", line 153, in _invoice_sale_orders
    invoices = confirmed_orders._create_invoices()
  File "/odoo/src/addons/sale_timesheet/models/sale_order.py", line 111, in _create_invoices
    moves = super()._create_invoices(grouped=grouped, final=final, date=date)
  File "/odoo/src/addons/sale/models/sale_order.py", line 775, in _create_invoices
    raise self._nothing_to_invoice_error()
odoo.exceptions.UserError: There is nothing to invoice!

Reason(s) of this behavior could be:
- You should deliver your products before invoicing them.
- You should modify the invoicing policy of your product: Open the product, go to the "Sales" tab and modify invoicing policy from "delivered quantities" to "ordered quantities". For Services, you should modify the Service Invoicing Policy to 'Prepaid'.


Also , attaching screenshot from front end:
Screenshot from 2023-04-24 20-26-58

**Comment by dhyeyp-emipro** [comment]
As per the logs, It looks like the product configuration-related issue.

**Comment by bvanjeelharia** [comment]
The changes seem to be fine

**Comment by bvanjeelharia** [comment]
@dhyeyp-emipro , Found these issue from logs, please check:
2023-05-05 08:10:53,217 70 ERROR odoodb odoo.addons.base.models.ir_cron: Call from cron Update Mollie Payments for server action #724 failed in Job #35 
Traceback (most recent call last):
  File "/odoo/src/odoo/models.py", line 5210, in ensure_one
    _id, = self._ids
ValueError: too many values to unpack (expected 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/odoo/src/odoo/tools/safe_eval.py", line 332, in safe_eval
    return unsafe_eval(c, globals_dict, locals_dict)
  File "", line 1, in <module>
  File "/odoo/external-src/mollie_subscription_ept/models/mollie_payment.py", line 149, in auto_update_payments
    if payment_obj and payment_obj.status != payment['status']:
  File "/odoo/src/odoo/fields.py", line 1075, in __get__
    record.ensure_one()
  File "/odoo/src/odoo/models.py", line 5213, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: mollie.payment(3166, 3167)

**Comment by dhyeyp-emipro** [comment]
@bvanjeelharia It is due to multiple payments associated with the same subscription. But didn't understand how multiple payments were associated with any single subscription.

File: mollie_subscription_ept/models/mollie_payment.py [hunk 12]
```diff
@@ -165,8 +165,8 @@ def create_mollie_invoice(self):
                 order = subscription.sale_order_id
                 if order.state == "draft":
                     order.action_confirm()
-                if not order.invoice_ids:
-                    invoice = order._create_invoices()
+                # if not order.invoice_ids:
```
Comment (bvanjeelharia) [inline mollie_subscription_ept/models/mollie_payment.py]: Please change this to 
<pre>
                if not self.env['ir.config_parameter'].sudo().get_param('sale.automatic_invoice',False):
                    if not order.invoice_ids:
                        invoice = order._create_invoices()
</pre>

As its not working in case , automatic invoicing is disabled

Comment (dhyeyp-emipro) [reply mollie_subscription_ept/models/mollie_payment.py]: PR Updated, Will you please verify it from your end?

**Review by bvanjeelharia — APPROVED** [review]

**Review by bvanjeelharia — APPROVED** [review]

**Comment by odoo-mvds** [comment]
Hi,

Thank you for your contribution and for submitting the pull request (PR).

We have reviewed your PR and appreciate your effort. Currently, we have no plans to implement a subscription module in the near future. When we do address this feature, we aim to keep the code minimal and manage it via tokens/mandates. As a result, we will be closing this PR.

We truly appreciate your effort and interest in improving the project.

Best regards,
